rustls/examples/internal/test_ca.rs

```diff
@@ -0,0 +1,321 @@
+                        role.label(),
+                        f.0
+                    )))?;
+                    for cert in f.1 {
+                        chain_file.write_all(cert.pem().as_bytes())?;
+                    }
+                }
+            }
+
+            certified_keys.insert((role, alg.inner), CertifiedKey { cert, key_pair });
+        }
+    }
+
+    Ok(())
+}
+
+fn crl_for_serial(serial_number: &SerialNumber) -> CertificateRevocationListParams {
+    let now = OffsetDateTime::now_utc();
+    CertificateRevocationListParams {
+        this_update: now,
+        next_update: now + Duration::from_secs(60 * 60 * 24 * 5),
+        crl_number: SerialNumber::from(1234),
+        issuing_distribution_point: None,
+        revoked_certs: vec![RevokedCertParams {
+            serial_number: serial_number.clone(),
+            revocation_time: now,
+            reason_code: Some(RevocationReason::KeyCompromise),
+            invalidity_date: None,
+        }],
+        key_identifier_method: KeyIdMethod::Sha256,
+    }
+}
+
+#[derive(Debug, Clone, Copy, Eq, PartialEq, Hash)]
+enum Role {
+    Client,
+    EndEntity,
+    Intermediate,
+    TrustAnchor,
+}
+
+impl Role {
+    fn key_file_path(&self, alg: &'static SigAlgContext) -> PathBuf {
+        alg.output_directory()
+            .join(format!("{}.key", self.label()))
+    }
+
+    fn cert_pem_file_path(&self, alg: &'static SigAlgContext) -> PathBuf {
+        alg.output_directory()
+            .join(format!("{}.cert", self.label()))
+    }
+
+    fn cert_der_file_path(&self, alg: &'static SigAlgContext) -> PathBuf {
+        alg.output_directory()
+            .join(format!("{}.der", self.label()))
+    }
+
+    fn label(&self) -> &'static str {
+        match self {
+            Self::Client => "client",
+            Self::EndEntity => "end",
+            Self::Intermediate => "inter",
+            Self::TrustAnchor => "ca",
+        }
+    }
+
+    fn params(&self, alg: &'static SigAlgContext) -> CertificateParams {
+        let mut params = CertificateParams::default();
+        params.distinguished_name = self.common_name(alg);
+        params.use_authority_key_identifier_extension = true;
+        let serial = SERIAL_NUMBER.fetch_add(1, Ordering::SeqCst);
+        params.serial_number = Some(SerialNumber::from_slice(&serial.to_be_bytes()[..]));
+
+        match self {
+            Self::TrustAnchor | Self::Intermediate => {
+                params.is_ca = IsCa::Ca(BasicConstraints::Unconstrained);
+                params.key_usages = ISSUER_KEY_USAGES.to_vec();
+                params.extended_key_usages = ISSUER_EXTENDED_KEY_USAGES.to_vec();
+            }
+            Self::EndEntity | Self::Client => {
+                params.is_ca = IsCa::NoCa;
+                params.key_usages = EE_KEY_USAGES.to_vec();
+                params.subject_alt_names = vec![
+                    SanType::DnsName(Ia5String::try_from("testserver.com".to_string()).unwrap()),
```

Comment:
Nit: where building a `String` from a `&str`, prefer `to_owned()` over `to_string()` (which is more general and more expensive).

Reply:
Done.